File: leetcode/53.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        currSum = 0
        maxSub = 0
        k = 0
        for n in nums : 
            currSum += n 
            if currSum < 0 :
                currSum = 0
                k+=1
            maxSub = max(maxSub, currSum)
        if k == len(nums) : 
            # in the case nums contains only negative numbers 
            return max(nums)
        return maxSub

        

File: leetcode/53.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int currSum = 0, maxSum = 0 ;
        int k = 0;

        for (int n : nums){
            currSum += n;
            if (currSum < 0){
                currSum = 0;
                k++ ;
            }
            maxSum = std::max(maxSum, currSum);    
        }
        if (k == nums.size()){
            return *std::max_element(nums.begin(), nums.end());
        }
        return maxSum;

    }
};
